source/opt/aggressive_dead_code_elim_pass.cpp

```diff
@@ -101,21 +101,33 @@ void AggressiveDCEPass::ProcessLoad(uint32_t varId) {
   live_local_vars_.insert(varId);
 }
 
-bool AggressiveDCEPass::IsStructuredIfHeader(ir::BasicBlock* bp,
-                                             ir::Instruction** mergeInst,
-                                             ir::Instruction** branchInst,
-                                             uint32_t* mergeBlockId) {
-  auto ii = bp->end();
-  --ii;
-  if (ii->opcode() != SpvOpBranchConditional) return false;
-  if (ii == bp->begin()) return false;
-  if (branchInst != nullptr) *branchInst = &*ii;
-  --ii;
-  if (ii->opcode() != SpvOpSelectionMerge) return false;
-  if (mergeInst != nullptr) *mergeInst = &*ii;
-  if (mergeBlockId != nullptr)
-    *mergeBlockId =
-        ii->GetSingleWordInOperand(kSelectionMergeMergeBlockIdInIdx);
+bool AggressiveDCEPass::IsStructuredHeader(ir::BasicBlock* bp, SpvOp mergeOp,
+                                           ir::Instruction** mergeInst,
+                                           ir::Instruction** branchInst,
+                                           uint32_t* mergeBlockId) {
+  ir::Instruction* tempMergeInst;
+  if (!IsStructuredIfOrLoopHeader(bp, &tempMergeInst, branchInst, mergeBlockId))
+    return false;
+  if (mergeInst != nullptr) *mergeInst = tempMergeInst;
+  return tempMergeInst->opcode() == mergeOp;
+}
+
+bool AggressiveDCEPass::IsStructuredIfOrLoopHeader(ir::BasicBlock* bp,
+                                                   ir::Instruction** mergeInst,
+                                                   ir::Instruction** branchInst,
+                                                   uint32_t* mergeBlockId) {
+  auto brii = bp->end();
```

Comment:
BasicBlock has some useful functionality in this regard. Pleas rewrite in terms of BasicBlock::GetMergeInst, BasicBlock::tail and BasicBlock::MergeIdIfAny.

Reply:
OK.